internal/repository/repository.go

```diff
@@ -515,6 +517,85 @@ func (r *Repository) LoadIndex(ctx context.Context) error {
 	return nil
 }
 
+const listPackParallelism = 10
+
+// LoadIndexFromPacks loads a new index by reading all given pack files
+// This will create one Index data structure per pack file read that is not
+// marked as finalized.
+func (r *Repository) LoadIndexFromPacks(ctx context.Context, packs map[restic.ID]int64, p *progress.Counter) (invalid restic.IDs, err error) {
+	var m sync.Mutex
+
+	debug.Log("Loading index from pack files")
+
+	// track spawned goroutines using wg, create a new context which is
+	// cancelled as soon as an error occurs.
+	wg, ctx := errgroup.WithContext(ctx)
+
+	type FileInfo struct {
+		restic.ID
+		Size int64
+	}
+	ch := make(chan FileInfo)
+	indexCh := make(chan *Index)
+
+	// send list of pack files through ch, which is closed afterwards
+	wg.Go(func() error {
+		defer close(ch)
+		for id, size := range packs {
+			select {
+			case <-ctx.Done():
+				return nil
+			case ch <- FileInfo{id, size}:
+			}
+		}
+		return nil
+	})
+
+	// a worker receives an pack ID from ch, reads the pack contents, and sends it to indexCh
+	worker := func() error {
+		for fi := range ch {
+			entries, _, err := r.ListPack(ctx, fi.ID, fi.Size)
+			if err != nil {
+				debug.Log("unable to list pack file %v", fi.ID.Str())
+				m.Lock()
+				invalid = append(invalid, fi.ID)
+				m.Unlock()
+			}
+			idx := NewIndex()
+			idx.StorePack(fi.ID, entries)
```

Comment:
Is there a specific reason to use individual indexes and not just one? Or one per worker thread? I'd expect a single index to use quite a bit less memory.

Reply:
Thanks for giving the hint. In fact, the only reason was that I copied the code from `LoadIndex`. Is now changed such that only one index is created.